core/embed/firmware/bl_check.c

```diff
@@ -174,12 +206,27 @@ void check_and_replace_bootloader(void) {
 
   ensure(flash_area_erase(&BOOTLOADER_AREA, NULL), NULL);
   ensure(flash_unlock_write(), NULL);
-  for (int i = 0; i < len / sizeof(uint32_t); i++) {
-    ensure(
-        flash_area_write_word(&BOOTLOADER_AREA, i * sizeof(uint32_t), data[i]),
-        NULL);
-  }
-  for (int i = len / sizeof(uint32_t); i < 128 * 1024 / sizeof(uint32_t); i++) {
+
+  uint32_t j = 0;
+  uint32_t d_len = 0;
+
+  do {
+    d_len = decomp.dest - decomp_out;
+    for (int i = 0; i < d_len / sizeof(uint32_t); i++) {
```

Comment:
If the length of the last decompressed block is not aligned to a 32-bit boundary, the last few bytes are left unwritten

Reply:
fixed https://github.com/trezor/trezor-firmware/pull/3304/commits/9143bf167745d860251791ae305470e98e153266